### ISSUE TITLE
Predicted result

### ISSUE THREAD

**Issue by ylyenjoy** [open-mmlab/mmaction2#410]
**Hello，I have finished my training with two classes,when I used the checkpoints,I got score like this:**
The top-5 labels with corresponding scores are:
normal:  0.21871684
turn:  -0.24892017
**or** 
The top-5 labels with corresponding scores are:
normal:  0.4902463
turn:  -0.52880824
**Do these results look normal？Is it not normalized for the output of the model？Thanks**

**Comment by innerlee** [comment]
Try change this in the config
```yaml
test_cfg = dict(average_clips='prob')
```

**Comment by ylyenjoy** [comment]
**Now I change my config as you say,but it doesn't work,My score still likes before:**
The top-5 labels with corresponding scores are:
turn:  0.34423405
normal:  -0.36384612

**Comment by innerlee** [comment]
Could you share your config and shell command?

**Comment by irvingzhang0512** [comment]
> **Now I change my config as you say,but it doesn't work,My score still likes before:**
> The top-5 labels with corresponding scores are:
> turn: 0.34423405
> normal: -0.36384612

which model? I guess tsn or tsm?
If you use latest codes, `test_cfg = dict(average_clips='prob')` should work.
Or you can try  `test_cfg = dict(average_clips='prob', test_crops=1)`

**Comment by ylyenjoy** [comment]
> 您可以共享您的config和shell命令吗？
OH,I set 'test_cfg = dict(average_clips='prob', test_crops=1)',now the score looks normal,Thank you
The top-5 labels with corresponding scores are:
normal:  0.73478836
turn:  0.26521158

**Comment by ylyenjoy** [comment]
> > **现在我按照您的说明更改配置，但是它不起作用，我的分数仍然像以前一样：**
> > 带有相应分数的前5个标签是：
> > 转弯：0.34423405
> > 正常：-0.36384612
> 
> 哪个型号？我猜是tsn还是tsm？
> 如果您使用最新代码，则`test_cfg = dict(average_clips='prob')`应该可以。
> 或者你可以尝试 `test_cfg = dict(average_clips='prob', test_crops=1)`

OK,I have tried  'test_cfg = dict(average_clips='prob', test_crops=1)'，it works,thank you

**Comment by dreamerlin** [comment]
It seems you should update your code according to master

**Comment by dreamerlin** [comment]
If you feel we have help you, give us a STAR!

**Comment by wwdok** [comment]
hi, @irvingzhang0512 , i just trained my TSM model with two action classes, the training config is :
```
Config:
model = dict(
    type='Recognizer2D',
    backbone=dict(
        type='ResNetTSM',
        pretrained='torchvision://resnet50',
        depth=50,
        norm_eval=False,
        shift_div=8),
    cls_head=dict(
        type='TSMHead',
        num_classes=2,
        in_channels=2048,
        spatial_type='avg',
        consensus=dict(type='AvgConsensus', dim=1),
        dropout_ratio=0.5,
        init_std=0.001,
        is_shift=True))
train_cfg = None
test_cfg = dict(average_clips='prob')
dataset_type = 'VideoDataset'
data_root = 'myactiondataset/train/'
data_root_val = 'myactiondataset/val/'
ann_file_train = 'myactiondataset/myactiondataset_train_video.txt'
ann_file_val = 'myactiondataset/myactiondataset_val_video.txt'
ann_file_test = 'myactiondataset/myactiondataset_test_video.txt'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_bgr=False)
train_pipeline = [
    dict(type='DecordInit'),
    dict(type='SampleFrames', clip_len=1, frame_interval=1, num_clips=8),
    dict(type='DecordDecode'),
    dict(type='Resize', scale=(-1, 256)),
    dict(
        type='MultiScaleCrop',
        input_size=224,
        scales=(1, 0.875, 0.75, 0.66),
        random_crop=False,
        max_wh_scale_gap=1,
        num_fixed_crops=13),
    dict(type='Resize', scale=(224, 224), keep_ratio=False),
    dict(type='Flip', flip_ratio=0.5),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_bgr=False),
    dict(type='FormatShape', input_format='NCHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs', 'label'])
]
val_pipeline = [
    dict(type='DecordInit'),
    dict(
        type='SampleFrames',
        clip_len=1,
        frame_interval=1,
        num_clips=8,
        test_mode=True),
    dict(type='DecordDecode'),
    dict(type='Resize', scale=(-1, 256)),
    dict(type='CenterCrop', crop_size=224),
    dict(type='Flip', flip_ratio=0),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_bgr=False),
    dict(type='FormatShape', input_format='NCHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
test_pipeline = [
    dict(type='DecordInit'),
    dict(
        type='SampleFrames',
        clip_len=1,
        frame_interval=1,
        num_clips=8,
        test_mode=True),
    dict(type='DecordDecode'),
    dict(type='Resize', scale=(-1, 256)),
    dict(type='CenterCrop', crop_size=224),
    dict(type='Flip', flip_ratio=0),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_bgr=False),
    dict(type='FormatShape', input_format='NCHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
data = dict(
    videos_per_gpu=8,
    workers_per_gpu=4,
    train=dict(
        type='VideoDataset',
        ann_file='myactiondataset/myactiondataset_train_video.txt',
        data_prefix='myactiondataset/train/',
        pipeline=[
            dict(type='DecordInit'),
            dict(
                type='SampleFrames', clip_len=1, frame_interval=1,
                num_clips=8),
            dict(type='DecordDecode'),
            dict(type='Resize', scale=(-1, 256)),
            dict(
                type='MultiScaleCrop',
                input_size=224,
                scales=(1, 0.875, 0.75, 0.66),
                random_crop=False,
                max_wh_scale_gap=1,
                num_fixed_crops=13),
            dict(type='Resize', scale=(224, 224), keep_ratio=False),
            dict(type='Flip', flip_ratio=0.5),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_bgr=False),
            dict(type='FormatShape', input_format='NCHW'),
            dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
            dict(type='ToTensor', keys=['imgs', 'label'])
        ]),
    val=dict(
        type='VideoDataset',
        ann_file='myactiondataset/myactiondataset_val_video.txt',
        data_prefix='myactiondataset/val/',
        pipeline=[
            dict(type='DecordInit'),
            dict(
                type='SampleFrames',
                clip_len=1,
                frame_interval=1,
                num_clips=8,
                test_mode=True),
            dict(type='DecordDecode'),
            dict(type='Resize', scale=(-1, 256)),
            dict(type='CenterCrop', crop_size=224),
            dict(type='Flip', flip_ratio=0),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_bgr=False),
            dict(type='FormatShape', input_format='NCHW'),
            dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
            dict(type='ToTensor', keys=['imgs'])
        ]),
    test=dict(
        type='VideoDataset',
        ann_file='myactiondataset/myactiondataset_test_video.txt',
        data_prefix='myactiondataset/val/',
        pipeline=[
            dict(type='DecordInit'),
            dict(
                type='SampleFrames',
                clip_len=1,
                frame_interval=1,
                num_clips=8,
                test_mode=True),
            dict(type='DecordDecode'),
            dict(type='Resize', scale=(-1, 256)),
            dict(type='CenterCrop', crop_size=224),
            dict(type='Flip', flip_ratio=0),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_bgr=False),
            dict(type='FormatShape', input_format='NCHW'),
            dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
            dict(type='ToTensor', keys=['imgs'])
        ]))
optimizer = dict(
    type='SGD',
    constructor='TSMOptimizerConstructor',
    paramwise_cfg=dict(fc_lr5=True),
    lr=0.00015625,
    momentum=0.9,
    weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=20, norm_type=2))
lr_config = dict(policy='step', step=[20, 40])
total_epochs = 30
checkpoint_config = dict(interval=10)
evaluation = dict(
    interval=5, metrics=['top_k_accuracy', 'mean_class_accuracy'])
log_config = dict(interval=5, hooks=[dict(type='TextLoggerHook')])
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = './tutorial_exps'
load_from = './checkpoints/tsm_r50_video_1x1x8_100e_kinetics400_rgb_20200702-a77f4328.pth'
resume_from = None
workflow = [('train', 1)]
omnisource = False
seed = 0
gpu_ids = range(0, 1)
```

, so the `test_cfg = dict(average_clips='prob')` is without `test_crops=1`, then the result output is like : 
```
[('fall_down', 1.1727579), ('smoking', -1.1995307)]
[('fall_down', 1.655934), ('smoking', -1.6735543)]
[('fall_down', 2.6001232), ('smoking', -2.6183808)]
```
It seems the `test_crops=1` makes the difference, could you please tell me where is the related source code ? thanks ! In the meantime, i am retaining model with adding `test_crops=1`. If i want to convert pth format to onnx format using pytorch2onnx.py, do i need to add `test_crops=1` in the config file too ?

**Comment by irvingzhang0512** [comment]
@wwdok which version of mmaction2 are you using now? `test_crops` is useless since v0.9.0. `test_cfg = dict(average_clips='prob')` is all you need.

https://github.com/open-mmlab/mmaction2/blob/a6da5441397bd05a3e57d34b69bfd72e846f9dac/mmaction/models/recognizers/recognizer2d.py#L35

**Comment by wwdok** [comment]
Thank you ! I use mmaction2==0.10.0 within colab, I check the training log, i found it is different from that without `test_crops=1`, If i remember correctly, previously the loss_loss and loss are almost 0.0x at 10th epoch
![image](https://user-images.githubusercontent.com/43233772/104867747-89f0ab80-597c-11eb-9a74-995c7b6b9f56.png)
maybe the problem occurs in the process of `pytorch2onnx.py` ? I need to add `test_crops=1` in its corresponding config file  ? I will try it later and update this comment ~

**Comment by irvingzhang0512** [comment]
Actually, `test_cfg` has no effect on training. v0.10.0 doesn't support `test_crops`, you don't need `test_crops` at all.

How did you get results like `[('fall_down', 1.1727579), ('smoking', -1.1995307)]`?

**Comment by wwdok** [comment]
Do you mean i don't need to add `test_crops=1` in the config file as well when i am using `pytorch2onnx.py` to convert pth format to onnx format ?
I firstly convert it to onnx format, then convert to tensorrt format, the image fed into tensorrt is preprocessed by following code:
```
    def preprocess_img(self, img):
        image = Image.fromarray(np.uint8(img))
        mean = [0.485, 0.456, 0.406]
        std = [0.229, 0.224, 0.225]
        transform = transforms.Compose([
            transforms.Resize((self.model_input_height, self.model_input_width)),
            transforms.ToTensor(),
            transforms.Normalize(mean, std, inplace=False),
        ])
        normalized_img = transform(image)
        img_np = normalized_img.numpy()
        img_np = img_np.ravel()
        return img_np
```

**Comment by irvingzhang0512** [comment]
> Do you mean i don't need to add `test_crops=1` in the config file as well when i am using `pytorch2onnx.py` to convert pth format to onnx format ?
> I firstly convert it to onnx format, then convert to tensorrt format, the image fed into tensorrt is preprocessed by following code:
> 
> ```
>     def preprocess_img(self, img):
>         image = Image.fromarray(np.uint8(img))
>         mean = [0.485, 0.456, 0.406]
>         std = [0.229, 0.224, 0.225]
>         transform = transforms.Compose([
>             transforms.Resize((self.model_input_height, self.model_input_width)),
>             transforms.ToTensor(),
>             transforms.Normalize(mean, std, inplace=False),
>         ])
>         normalized_img = transform(image)
>         img_np = normalized_img.numpy()
>         img_np = img_np.ravel()
>         return img_np
> ```

Yes, you don't need `test_crops=1` anywhere.

**Comment by wwdok** [comment]
I have run my trained pth model using `long_video_demo.py`, output video shows normal score scope (0, 1), but using its tensorrt version, it will output negative scores and plus one scores, i have check the two frames that fed into pth model and tensorrt model :
```
* first frame that fed into pth model :
[[[-2.1179, -2.1179, -2.1179,  ..., -2.0837, -2.0837, -2.0837],
          [-2.1179, -2.1179, -2.1179,  ..., -2.0837, -2.0837, -2.0837],
          [-2.0837, -2.1179, -2.1179,  ..., -2.1008, -2.1008, -2.1008],
          ...,
          [-1.9980, -1.9809, -1.9809,  ..., -1.0390, -1.0562, -1.0733],
          [-2.0152, -1.9980, -2.0152,  ..., -1.3644, -1.3815, -1.4158],
          [-2.0665, -2.0494, -2.0494,  ..., -1.6555, -1.6727, -1.6898]],

         [[-2.0357, -2.0357, -2.0357,  ..., -2.0007, -2.0007, -2.0007],
          [-2.0357, -2.0182, -2.0357,  ..., -2.0007, -2.0007, -2.0007],
          [-2.0007, -2.0357, -2.0182,  ..., -2.0182, -2.0182, -2.0182],
          ...,
          [-1.8081, -1.7731, -1.7731,  ..., -0.9153, -0.9153, -0.8978],
          [-1.8256, -1.7906, -1.7906,  ..., -1.2654, -1.2479, -1.2479],
          [-1.8782, -1.8431, -1.8431,  ..., -1.5630, -1.5455, -1.5455]],

         [[-1.8044, -1.8044, -1.8044,  ..., -1.8044, -1.8044, -1.8044],
          [-1.8044, -1.7870, -1.8044,  ..., -1.8044, -1.8044, -1.8044],
          [-1.7696, -1.8044, -1.7870,  ..., -1.8044, -1.8044, -1.8044],
          ...,
          [-1.6476, -1.6302, -1.6302,  ..., -0.6018, -0.6018, -0.6018],
          [-1.6824, -1.6476, -1.6476,  ..., -0.9504, -0.9504, -0.9504],
          [-1.7173, -1.6999, -1.6999,  ..., -1.2293, -1.2293, -1.2293]]]

* first frame that fed into tensorrt model:
[-2.117904  -2.117904  -2.117904  ... -1.316427  -1.3338562 -1.3512855]
```
we can see that the first three values are all -2.1179, but the last three values are little different.

In the meanwhile, i am thinking if it may be that the tensorrt model did not do post-processing before it output score ?

If you are interested in my issue and have time, i could share you the code and tensorrt model to reproduce this error :)

**Comment by irvingzhang0512** [comment]
Maybe you could open a new issue about this. @wwdok 

I'm not familar with onnx. Maybe the last `F.softmax` is not exported to onnx model. You could visualize onnx model to find out.

**Comment by wwdok** [comment]
@irvingzhang0512 I have open a issue [here](https://github.com/open-mmlab/mmaction2/issues/552), thanks for your discussion !
